lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+    SmallVector<SILValue, 8> newArgs;
+    if (!stayInLoop) {
+      for (unsigned i = 0; i < oldHeaderNumArgs; ++i) {
+        newArgs.push_back(newHeader->getArgument(i));
+      }
+    }
+    SILBuilder builder(src->getTerminator());
+    SILLocation location(
+        getUserSourceLocation(src->getTerminator()->getDebugLocation()));
+    // Find an appropriate value to use for each escaping value.
+    unsigned argIndex = oldHeaderNumArgs;
+    for (const SILValue &escapingValue : escapingValues) {
+      if (DI.properlyDominates(escapingValue, src->getTerminator())) {
+        newArgs.push_back(escapingValue);
+      } else if (src != preheader) {
+        // if src is not the preheader than new header arguments are available.
+        newArgs.push_back(newHeader->getArgument(argIndex));
+      } else {
+        newArgs.push_back(getUndef(escapingValue->getType()));
+      }
+      ++argIndex;
+    }
+    // `exitIndex` to identify the block to which we exit from the loop.
+    // (insert a new value or get the old key value pair.)
+    auto emplaceResult = exitIndices.try_emplace(tgt, nextExitIndex);
+    if (emplaceResult.second) {
+      // Increment index as we inserted a new entry into the table.
+      ++nextExitIndex;
+    }
+    auto kvPair = *(emplaceResult.first);
+    newArgs.push_back(createTFIntegerConst(builder, location,
+                                           /*bitwidth*/ 32,
+                                           /*exitIndex*/ kvPair.second));
+    // `stayInLoop` flag
+    newArgs.push_back(createTFIntegerConst(builder, location,
+                                           /*bitwidth*/ 1, stayInLoop));
+    appendArguments(src->getTerminator(), newTgt, newArgs);
+  }
+
+  // Create a new exit block.
+  auto createBlockOutsideLoop = [this, loop, &currentFn]() {
+    SILBasicBlock *newBlock = currentFn->createBasicBlock();
+    SILLoop *parentLoop = loop->getParentLoop();
+    if (parentLoop) {
+      parentLoop->addBasicBlockToLoop(newBlock, LI.getBase());
+    }
+    return newBlock;
+  };
+
+  // FIXME: We can avoid creating an additional block and instead connect the
+  // header directly to the demuxBlock created in the loop below. Alternatively,
+  // we can also use contractUncondBranches in TFParititon.cpp to remove this
+  // block later.
+  SILBasicBlock *newExitBlock = createBlockOutsideLoop();
+  // Build a demuxing if..then..else block that can be used as a single
+  // exit block. This will not have any effect if there is a single exit
+  // block already.
+  auto curBlockIter = exitBlocks.rbegin();
+  SILBasicBlock *demuxBlock = *curBlockIter++;
+  SILLocation headerLocation =
+      getUserSourceLocation(header->getTerminator()->getDebugLocation());
+
+  while (curBlockIter != exitBlocks.rend()) {
+    SILBasicBlock *newBlock = createBlockOutsideLoop();
+    SILBasicBlock *trueBlock = *curBlockIter++;
+    builder.setInsertionPoint(newBlock);
+    SILValue condTensor = builder.createBuiltin(
```

Comment:
Would it make sense to use `graph_op` instruction instead? We are moving towards `graph_op` for all code paths after deabstraction.

Reply:
Done.